api/security_firewall.go

```diff
@@ -100,30 +103,31 @@ func (api *API) updateFirewallSettingsWithRetry(instanceID int, params []map[str
 	log.Printf("[DEBUG] go-api::security_firewall::update instance id: %v, params: %v", instanceID, params)
 	path := fmt.Sprintf("/api/instances/%d/security/firewall", instanceID)
 	response, err := api.sling.New().Put(path).BodyJSON(params).Receive(nil, &failed)
-
 	if err != nil {
 		return err
 	}
 
-	statusCode := response.StatusCode
-	log.Printf("[DEBUG] go-api::security_firewall::update statusCode: %d", statusCode)
 	switch {
-	case statusCode == 400:
-		if strings.Compare(failed["message"].(string), "Your new firewall rules have not finished "+
-			"configuring yet,try again in a few minutes") == 0 {
+	case response.StatusCode == 204:
+		return nil
+	case response.StatusCode == 400:
+		switch {
+		case failed["error_code"] == nil:
+			break
+		case failed["error_code"].(float64) == 40001:
 			if attempts--; attempts > 0 {
 				log.Printf("[INFO] go-api::security_firewall::update Firewall not finished configuring "+
 					"attempts left %d and retry in %d seconds", attempts, sleep)
 				time.Sleep(time.Duration(sleep) * time.Second)
 				return api.updateFirewallSettingsWithRetry(instanceID, params, attempts, sleep)
 			} else {
-				return fmt.Errorf("Update firewall rules failed, status: %v, message: %s", response.StatusCode, failed)
+				break
 			}
+		case failed["error_code"].(float64) == 40002:
+			return fmt.Errorf("Validation failed, invalid address")
```

Comment:
Not sure this is fully correct Go, but I want the message to be more clear, and also that we pass along the error message from the API. The validation error can happen for several things (invalid service name, invalid rule combo, invalid address)

```suggestion
			return fmt.Errorf("Firewall rules validation failed due to: %s", failed["error"])
```